Develop stream 2025-01-23

Snektron opened the pull request:
This pull request is based on top of https://github.com/ROCm/rocPRIM/pull/674, that one should be merged first.

This pull request contains our changes from the past few months. A summary of the most notable changes:
- Changes required for updating hipCUB and rocThrust to CCCL 2.6.0.
- Add rocPRIM versions of `UnrolledThreadLoad` and `UnrolledThreadCopy`
- Device histogram workaround for dynamic shared memory not working on Windows
- Removed `ROCPRIM_TARGET_ARCH`, which allowed users to specialize rocPRIM for a particular architecture. This is now mostly resolved by the new `ROCPRIM_TARGET_(CDNA3|CDNA2|CDNA1|RDNA3|RDNA2|RNDA1|GCN5)` macros, which are automatically determined from the (device) compilation target. This does not affect any downstream user code, but does affect rocThrust a little bit.
- Undeprecated and improved `rocprim::thread_load` and `rocprim::thread_store`. Most of the inline assembly has been removed, which should make the compiler able to optimize these operations better. These functions allow programmers to load memory more efficiently by providing hints for how memory will be used, for example, `rocprim::thread_load<rocprim::load_nontemporal /* previously: rocprim::load_cs */>(ptr)` can be used effectively to improve memory throughput on MI300.
- Added `rocprim::merge_inplace`. This algorithm is like `rocprim::merge`, but (as the name suggests) can operate in-place. While `rocprim::merge` is usually faster, this algorithm can be used when not enough memory is available for an out-of-place buffer.
- Adopt visibility rules to avoid linking problems when linking with separate versions of rocPRIM.
- Add 128-bit integer benchmarking and tuning to all algorithms.

jblok27 committed:
Fix tests triggering static asserts

jblok27 committed:
Fix static asserts and remove template specialization

jblok27 committed:
Fix benchmarks triggering static asserts

jblok27 committed:
Set explicit return types

Beanavil committed:
Resolve "Add hipGraph compatibility for device run length encode"

Beanavil committed:
Resolve "Investigate RLE (non trivial runs) failure"

Beanavil committed:
Resolve "Investigate RLE (non trivial runs) failure"

jblok27 committed:
Revert "Fix static asserts and remove template specialization"

This reverts commit b2089eeedec6b32415f27d0e3c0860442b0ba5e0.

jblok27 committed:
Revert "Fix benchmarks triggering static asserts"

This reverts commit a1634a9382e268285c1e21c0743b6a98e67a69ef.

jblok27 committed:
Revert "Fix tests triggering static asserts"

This reverts commit 24a36cb86283098036a5931562d0b4d3b42ca675.

jblok27 committed:
Fix formatting and copyright date

NB4444 committed:
Temp fix: histogram dynamic memory failure on windows

parbenc committed:
add ROCPRIM_ENABLE_ASSERTS to cmake to enable asserts in release build
set ROCPRIM_ENABLE_ASSERTS for CI tests

MyNameIsTrez committed:
Add missing docs for `block_run_length_decode`

yungshengtu committed:
Resolve "Config tuning and dynamic dispatch for device run length encode"

parbenc committed:
Resolve "Add documentation and test for out-of-range behaviour on merge path search"

add test for out of bounds diagonal test for rocprim::merge_path_search
document behaviour for out of bounds diagonal test for rocprim::merge_path_search

cenxuantian committed:
Resolve "Replace __launch_bounds__ with ROCPRIM_LAUNCH_BOUNDS in rocPRIM API"

cenxuantian committed:
Resolve "Implement type traits interface for rocPRIM's compiler-extended types and third-party's types"

Snektron committed:
remove ordered block id from run length encode

A missing call to __syncthreads() caused this algorithm to occasionally
produce an invalid index and cause a crash. Instead of just adding that
call, we can remove the use of ordered block id by the regular block id,
which is already ordered per HSA guarantees.

There still remains an issue with a hang, but this seems a runtime issue
and will be investigated separately.

Beanavil committed:
Update rmake.py copyright

Naraenda committed:
fix double buffers in device radix sort

cenxuantian committed:
Resolve "Replace hipMalloc, hipMallocHelper, and hipFree in tests with device_ptr"

ex-rzr committed:
fix device adjacent difference tuning benchmark

Beanavil committed:
Fix warp dpp reduction

ex-rzr committed:
Fix static assertion for int128 in ONNX Runtime with GNU extensions

Snektron committed:
fix GenerateResourceSpec for gfx90a

parbenc committed:
Resolve "ROCPRIM_TARGET_ARCH deprecation"

MyNameIsTrez committed:
Fix warnings when generating sphinx docs

MiloLurati committed:
Performance improvements with shuffled warp exchange

Naraenda committed:
log more information during configure

MyNameIsTrez committed:
Update documentation generation instructions

cenxuantian committed:
Resolve "Replace hipMalloc, hipMallocHelper, and hipFree in benchmarks and tests with device_ptr" Part 4

cenxuantian committed:
Resolve "search_n optimization"

MyNameIsTrez committed:
Remove select_arch and select_arch_case

NB4444 committed:
CI changes to speedup development

Naraenda committed:
fix missing equality operator and function constness

MyNameIsTrez committed:
Print arguments and timing info in `batch_memcpy_func` when `debug_synchronous` is true

cenxuantian committed:
Resolve "Replace hipMalloc, hipMallocHelper, and hipFree in benchmarks and tests with device_ptr" Part 3

cenxuantian committed:
Resolve "Replace hipMalloc, hipMallocHelper, and hipFree in benchmarks and tests with device_ptr" Part 1

cenxuantian committed:
Resolve "Replace hipMalloc, hipMallocHelper, and hipFree in benchmarks and tests with device_ptr" Part 2

NB4444 committed:
Fix rules overwrite for other then MR pipelines

NB4444 committed:
Remove windows CI check

Snektron committed:
MI200 & MI300 tuning

cenxuantian committed:
Resolve "Add stream, memset, array, and unique_ptr support to device_ptr"

NB4444 committed:
Resolve "Reduce amount of test runs based on environment variable."

Saiyang-Zhang committed:
Resolve "Get clarity on skipping of adjacent difference tests"

NB4444 committed:
Resolve "Add UnrolledThreadLoad and UnrolledThreadCopy functions"

NB4444 committed:
Resolve "General changes CCCL 2.6"

Snektron committed:
Reduce CI artifact retention time

NB4444 committed:
Resolve "Add thread_load_volatile_pointer to rocprim"

ex-rzr committed:
Fix test:install on rocm/dev-ubuntu

Saiyang-Zhang committed:
Resolve "Nth element - Investigate smaller index sizes where possible"

ex-rzr committed:
Enable int128 tests on Windows

yungshengtu committed:
Resolve "Benchmark and tuning for 128-bit integers"

cenxuantian committed:
Resolve "Verify sorting on device"

cenxuantian committed:
Resolve "Expose lookback_scan_state::create and lookback_scan_state::get_storage_size as ROCPRIM_HOST_DEVICE."

NB4444 committed:
To remove error in CI change it to AMDGPU_TARGETS

Naraenda committed:
add naive in-place api for device merge

NB4444 committed:
CI skip label fix

Naraenda committed:
fix device_ptr assign move and void types

Beanavil committed:
Resolve "Add block- and warp-level inclusive_scan with initial value support"

MyNameIsTrez committed:
Fix C++ includes showing up in the docs

NB4444 committed:
Resolve "Performance regressions during tuning: Adjacent find"

Naraenda committed:
fix doxygen warning in merge.rst

EnricoDeg committed:
Resolve "Autotuning for segmented reduce"

NB4444 committed:
Resolve "Update thread_store for parity"

Snektron committed:
Optimizations for MI300 unique performance

cenxuantian committed:
Resolve "Fix compile errors in "rocprim-release-tune" mode"

borysborys committed:
Documentation to CI upload

umfranzw committed:
Avoid evaluating CMake LINK_LANGUAGE generator expressions in CMake <3.18 (#667)

Recently, we added CMake HIP Language support. That change introduced generator expressions
that use the LINK_LANGUAGE keyword. This keyword is only supported in CMake 3.18+.
The minimum CMake version we require is 3.16.

As a result, if you built with CMake 3.16 - 3.18, you'd get a build failure.

This change adds checks around all the spots where the LINK_LANGUAGE keyword is used
to avoid this problem.

It also adds a description of the CMake HIP Language support to the changelog (previously overlooked).

NB4444 committed:
Resolve "Remove ASM from thread_load and thread_store"

MyNameIsTrez committed:
Add documentation steps to developer guidelines

cenxuantian committed:
Resolve "Refactor lookback_scan_state API"

ex-rzr committed:
Adopt symbol visibility rules in rocPRIM

MyNameIsTrez committed:
Remove extra dashes in the documentation its \param and \tparam

Snektron committed:
remove nm debug messages from cmake

Snektron committed:
improve type traits interface docs

umfranzw committed:
Fix Windows CMake link issue (#675)

A recent cmake change added a condition check that avoids evaluating
a LINK_LANGUAGE generator expression if HIP_USECXX is OFF (since
LINK_LANGUAGE is unsupported in CMake < 3.18). However, it didn't
account for the fact that the LINK_LANGUAGE expression is embedded in a
conditional generator expression.

This change adds an "else" block that ensures the "false" side of the
conditional generator expression is still evaluated when the LINK_LANGUAGE
code is skipped. This fixes some build issues on Windows.

Also updates the readme to include recent changes to the way projects
must link to rocPRIM.

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,42 @@
+
+### Added
+
+* Added missing operator `rocprim::key_value_pair::operator==`.
```
```suggestion
* Added `rocprim::key_value_pair::operator==`.
```
the fact that it was added implied that it was missing :)

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,42 @@
+### Added
+
+* Added missing operator `rocprim::key_value_pair::operator==`.
+* Added `rocprim::unrolled_copy` thread function to copy multiple items inside the thread.
```
```suggestion
* Added the `rocprim::unrolled_copy` function to copy multiple items inside a thread.
```

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,42 @@
+
+* Added missing operator `rocprim::key_value_pair::operator==`.
+* Added `rocprim::unrolled_copy` thread function to copy multiple items inside the thread.
+* Added `rocprim::unrolled_thread_load` thread function to load multiple items inside the thread using `rocprim::thread_load`.
```
```suggestion
* Added the `rocprim::unrolled_thread_load` function to load multiple items inside a thread using `rocprim::thread_load`.
```

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,42 @@
+* Added `rocprim::unrolled_copy` thread function to copy multiple items inside the thread.
+* Added `rocprim::unrolled_thread_load` thread function to load multiple items inside the thread using `rocprim::thread_load`.
+* Added `rocprim::int128_t` and `rocprim::uint128_t` to benchmarks for improved performance evaluation on 128-bit integers.
+* Added `rocprim::int128_t` to autotuning supported types to improve performance for 128-bit integers.
```
```suggestion
* Added `rocprim::int128_t` to the supported autotuning types to improve performance for 128-bit integers.
```

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,42 @@
+* Added `rocprim::unrolled_thread_load` thread function to load multiple items inside the thread using `rocprim::thread_load`.
+* Added `rocprim::int128_t` and `rocprim::uint128_t` to benchmarks for improved performance evaluation on 128-bit integers.
+* Added `rocprim::int128_t` to autotuning supported types to improve performance for 128-bit integers.
+* Added `rocprim::merge_inplace` function that allows merging in-place.
```
```suggestion
* Added the `rocprim::merge_inplace` function for merging in-place.
```

We want to avoid using "allow" as much as possible since it implies granting permission rather than providing and additional feature.

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,42 @@
+
+### Changed
+
+* Undeprecated `rocprim::thread_load` uses dereference as default and updated other cache load modifiers.
```
Was the function undeprecated, or a was a function with the same name added that does the same thing as the deprecated function, but with different parameters?

If that's the case, I suggest:

```suggestion
* A new version of `rocprim::thread_load replaces the deprecated `rocprim::thread_load` function. The new version of `thread_load` uses dereference by default and updates other cache load modifiers 
```

Also there needs to be clarity around "other cache load modifiers"; "other" as compared to what?

Snektron replied:
To give some background, these functions previously used inline assembly to emit special instructions for specialized load/store methods, for example, bypassing cache, or caching only in some cache levels. While this is useful, the inline assembly hindered optimization a lot, and as a result, using this function was almost always detrimental to performance. They were deprecated as a result. MI300 benefits from some of these methods, so we've improved the implementation to avoid inline assembly where possible. We plan to use these functions inside rocPRIM to improve performance in some locations, but that has not been implemented yet.

Snektron replied:
I've changed the description of both `rocprim::thread_store` and `rocprim::thread_load`, and I've added a better description of the updates to the cache load modifiers.

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,42 @@
+### Changed
+
+* Undeprecated `rocprim::thread_load` uses dereference as default and updated other cache load modifiers.
+* Undeprecated `rocprim::thread_store` uses dereference as default and updated other cache store modifiers.
```
Similar question as above regarding whether this is replacing a deprecated version of the function, and what "other" is in relation to.

```suggestion
* A new version of `rocprim::thread_store` replaces the deprecated `rocprim::thread_store`. The new version of `thread_store` uses dereference by default and updates other cache store modifiers.
```

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,42 @@
+* Undeprecated `rocprim::thread_load` uses dereference as default and updated other cache load modifiers.
+* Undeprecated `rocprim::thread_store` uses dereference as default and updated other cache store modifiers.
+* Changed symbol visibility: all kernels now have hidden visibility, inline namespace with the library version is added, e.g. `rocprim::ROCPRIM_300400_NS::...`. This allows to link multiple libraries built with different versions of rocPRIM.
+
```
```suggestion
* All kernels now have hidden symbol visibility. All symbols now have inline namespaces that include the library version, for example, `rocprim::ROCPRIM_300400_NS::symbol` instead of `rocPRIM::symbol`, letting the user link multiple libraries built with different versions of rocPRIM.
```
This is aligning it with the language used in hipcub

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,42 @@
+
+### Added
+
+* Added hipGraph support to device run-length-encode for non trivial runs (`rocprim::run_length_encode_non_trivial_runs`).
```
```suggestion
* Added hipGraph support to device run-length-encode for nontrivial runs (`rocprim::run_length_encode_non_trivial_runs`).
```

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,42 @@
+### Added
+
+* Added hipGraph support to device run-length-encode for non trivial runs (`rocprim::run_length_encode_non_trivial_runs`).
+* Added configuration autotuning to device run-length-encode for non trivial runs (`rocprim::run_length_encode_non_trivial_runs`) for improved performance on selected architectures.
```
```suggestion
* Added configuration autotuning to device run-length-encode for nontrivial runs (`rocprim::run_length_encode_non_trivial_runs`) for improved performance on selected architectures.
```

spolifroni-amd reviewed (changes requested):
Small changes needed to the changelog.

Snektron commented:
Thanks for the quick review, @spolifroni-amd, your work is appreciated :) I'll update the pull request tomorrow.

Snektron committed:
update changelog

Snektron commented:
Applied the suggestions

Naraenda committed:
fix(test_device_merge_inplace.cpp): prevent merge inplace tests to tests to grow too big

ex-rzr committed:
Fix test_linking: install the test's libraries too

umfranzw commented:
There's a CI test failure for test_device_merge_inplace on gfx942. I'm able to reproduce it, and I'm trying to narrow things down. It looks like the failure happens for the largest test size used in the DeviceMergeInPlace test case: 
`DeviceMergeInplaceParams<int32_t,
                             random_data_generator<int32_t, 2, 4>,
                             random_data_generator<int32_t, 2, 4>,
                             large_sizes>`

Naraenda commented:
@umfranzw This might be part of the hard memory usage limiter on the CI. IIRC, it will watch for processess that use more than a certain amount of memory and kill those off. If we know the limit we can set the limit defined in ae4709fe4354607a0e34407332a40b3dbc31d1a9, better.

umfranzw commented:
> @umfranzw This might be part of the hard memory usage limiter on the CI. IIRC, it will watch for processess that use more than a certain amount of memory and kill those off. If we know the limit we can set the limit defined in [ae4709f](https://github.com/ROCm/rocPRIM/commit/ae4709fe4354607a0e34407332a40b3dbc31d1a9), better.

@Naraenda I ran some experiments, and it looks like if I allocate 2^35 total bytes, everything runs fine, but if I allocate 2^36 the process gets killed. I believe if you modify your existing check in https://github.com/ROCm/rocPRIM/commit/ae4709fe4354607a0e34407332a40b3dbc31d1a9 to look at `total_bytes` instead of `size_total`, it should work fine.

umfranzw commented on example/example_type_traits_interface.cpp:
```diff
@@ -0,0 +1,125 @@
+        = rocprim::traits::integral_sign::values<traits::integral_sign::kind::signed_type>;
+};
+
+int main()
```
This is just a minor nit-picky thing, but since main is declared with return type int, it might be nice to add a return statement at the end of the function so the compiler doesn't issue a warning.

Naraenda replied:
[The CPP reference does note that return statements are not required (`return 0;` is implicit).](https://en.cppreference.com/w/cpp/language/main_function) I have added it in a5361e7244777a602f39dfb16c40ca17424d2515 as there's not much harm in being more explicit.

umfranzw commented on rmake.py:
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/python3
-""" Copyright (c) 2021-2023 Advanced Micro Devices, Inc. All rights reserved.
+""" Copyright (c) 2021-2024 Advanced Micro Devices, Inc. All rights reserved.
```
The end year here should be 2025.

Naraenda replied:
Fixed in 84b078d2a81632924fe8bf50f1a5f10797311312

umfranzw reviewed (commented):
I think this looks good - I just found a few small things I wanted to mention.

Naraenda committed:
chore(rmake.py): bump copyright

Naraenda committed:
fix(example_type_traits_interface.cpp): explicit return from main

Naraenda commented:
> @Naraenda I ran some experiments, and it looks like if I allocate 2^35 total bytes, everything runs fine, but if I allocate 2^36 the process gets killed. I believe if you modify your existing check in [ae4709f](https://github.com/ROCm/rocPRIM/commit/ae4709fe4354607a0e34407332a40b3dbc31d1a9) to look at `total_bytes` instead of `size_total`, it should work fine.

I've updated the commit that introduced this: 46a1f80e7475d307b3ce98674f3477cacb973d33

umfranzw reviewed (approved):
I think this looks ready to go.

NB4444 commented:
I resolved the merge conflicts.

ex-rzr committed:
Fix unintentionally exported shared libs of test_linking

rocm_install adds all libraries to ${PROJECT_NAME_LOWER}-targets
(i.e. rocprim-targets.cmake) unless a custom export file is specified.
